Merge `react-native-membrane-webrtc` into `react-native-client-sdk` (#49)
fishjam-dev/react-native-client-sdk: README.md

Author: I copied file from `react-native-membrane-webrtc` (as I should do from the beginning)

Author: yes

# react-native-client-sdk

> react-native-client-sdk is a React Native wrapper for [membrane-webrtc-android](https://github.com/jellyfish-dev/membrane-webrtc-android) and [membrane-webrtc-ios](https://github.com/jellyfish-dev/membrane-webrtc-ios). It allows you to quickly and easily create a mobile client app in React Native for Membrane server.

Author: I updated Jellyfish reference.
But I will leave links to archived repos, as this library still depends on archived packages. We have to update `rn-client-sdk` dependencies to new packages in future.

```

### Expo plugin

> Even though this is not a native library, it has a dependency on `@jellyfish-dev/react-native-membrane-webrtc` which requires native configuration.

Author: Good catch. I thought I spotted all places with `react-native-membrane-webrtc`.

   Replace `{{GROUP_IDENTIFIER}}` and `{{BUNDLE_IDENTIFIER}}` with your group identifier and bundle identifier respectively.
9. Rebuild the app and enjoy!

# Example

We strongly recommend checking out our example app that implements a basic video room client. To run the app:

> 1. Go to Membrane's server demo repo: https://github.com/membraneframework/membrane_videoroom. Follow instructions there to setup and run demo server.

Author: I will leave this as follow up tasks - but this part should be re-written to use example from `js-server-sdk`

2. Clone the repo
3. ```
   $ cd example
   $ yarn
   ```
4. In App.ts replace server url with your server's url.
5. `yarn run android` or `yarn run ios` or run project from Android Studio / Xcode just like every RN project. Note that simulators won't work, you have to test on real device for camera and screensharing to run.

# Usage

> Important note!! Since version 7.4.0 call function `initializeWebRTC()` once in your app before using any other functionality.

Reviewer: ```suggestion
> [!IMPORTANT]  
> Since version 7.4.0, you need to call function `initializeWebRTC()` once in your app before using any other functionality.
```

      endpointMetadata: {
        displayName: 'Annie',
      },
      connectionParams: {
        token: 'TOKEN',
      },
    });
  } catch (e) {
    console.log('error!');
  }
};
```

Remember to gracefully disconnect from the server using the `disconnect()` function:

```ts
const stopServerConnection = () => {
  await disconnect();
};
```

Also handle errors properly, for example when internet connection fails or server is down:

```ts
useEffect(() => {
  if (error) console.log('error: ', e);
}, [error]);
```

Start the device's camera and microphone using `useCamera()` and `useMicrophone()` hooks. Use `videoTrackMetadata` and `audioTrackMetadata` options to send metadata about the tracks (for example whether it's a camera or screencast track).

```ts
const { startCamera } = useCamera();
const { startMicrophone } = useMicrophone();

await startCamera({
  quality: VideoQuality.HD_169,
  videoTrackMetadata: { active: true, type: 'camera' },
});
await startMicrophone({ audioTrackMetadata: { active: true, type: 'audio' } });
```

For more options and functions to control the camera and microphone see the API documentation.

> If you have the connection set up, then use `useEndpoints()` hook to track the other endpoints in the room. One of the endpoints will be a local participant (the one who's using the device). When endpoints is added or removed because an user joins or leaves the room, the endpoints will be updated automatically. Simply call the hook like this:

Reviewer: Please set up line wrapping (eg. [like this](https://github.com/jellyfish-dev/js-server-sdk/blob/137dcd4196a36216041855f8a6bc66ca7bbf4807/.prettierrc#L4)) and please make sure prettier will format `.md` files

Author: Added as follow up task

```ts
const endpoints = useEndpoints();
```

When you have the endpoints all that's left is to render their video tracks. Use `<VideoRendererView />` component like this:

```ts
{
  endpoint.videoTracks.map((track) => (
    <VideoRendererView trackId={track.id} />
  ));
}
```

You can style the views to lay out them however you'd like, basic animations should work too.

There are also some simple hooks for toggling camera, microphone and screensharing. Use them like this:

```ts
const { isCameraOn, toggleCamera } = useCameraState();
const { isMicrophoneOn, toggleMicrophone } = useMicrophoneState();
```

For screencasting use `useScreencast()` hook. The local endpoint will have a new video track which you can render just like an ordinary video track with <VideoRendererView />:

```ts
const { isScreencastOn, toggleScreencast } = useScreencast();
...
toggleScreencast({screencastMetadata: { displayName: "Annie's desktop" }});
```

Use track metadata to differentiate between video and screencast tracks.

### Developing

Run `./scripts/init.sh` in the main directory to install swift-format and set up git hooks.

> To release a new version of the lib just run `yarn release`, follow the prompts to bump version, make tags, commits and upload to npm

Reviewer: ```suggestion
To release a new version of the lib just run `yarn release`, follow the prompts to bump version, make tags, commits and upload to npm.
```

To release a new version of the example app on Android install fastlane, get upload key password and firebase auth json from the devs, update `~/.gradle/gradle.properties` like this:

```
MEMBRANE_UPLOAD_STORE_FILE=my-upload-key.keystore
MEMBRANE_UPLOAD_KEY_ALIAS=my-key-alias
MEMBRANE_UPLOAD_STORE_PASSWORD=********
MEMBRANE_UPLOAD_KEY_PASSWORD=********
```

and run `yarn releaseAppAndroid` from the main directory.

To release a new version of the example app on iOS install fastlane, get added to swmansion app store account and run `yarn releaseAppIos` from the main directory.

> Pro tip: when developing set backend url in `.env.development`.

Reviewer: ```suggestion
Pro tip: when developing, set the backend url in `.env.development`.
```
